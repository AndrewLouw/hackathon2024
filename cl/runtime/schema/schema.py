# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

import importlib
import inspect
from cl.runtime.records.class_info import ClassInfo
from cl.runtime.schema.type_decl import TypeDecl
from cl.runtime.schema.type_decl import pascalize
from cl.runtime.schema.type_decl_key import TypeDeclKey
from cl.runtime.settings.context_settings import ContextSettings
from collections import Counter
from collections import defaultdict
from enum import Enum
from memoization import cached
from pkgutil import walk_packages
from types import ModuleType
from typing import Dict
from typing import Iterable
from typing import List
from typing import Type
from typing_extensions import Self


def is_key_or_record(data_type):
    """Return true if the type is a record based on the presence of 'get_key_type' method."""
    return (
        inspect.isclass(data_type)
        and hasattr(data_type, "get_key_type")
        and not inspect.isabstract(data_type)
        and not data_type.__name__.endswith("Mixin")
    )


class Schema:
    """
    Provide declarations for the specified type and all dependencies.
    """

    _type_dict_by_short_name: Dict[str, Type] = None

    @classmethod
    @cached
    def get_types(cls) -> Iterable[Type]:
        """
        Get all record and field types found in the list of packages specified in settings.

        Notes:
            The result is returned in the alphabetical order of module.ClassName.
        """
        return cls.get_type_dict().values()

    @classmethod
    def add_types_to_dict_by_short_name(cls, types: Iterable[Type]) -> None:
        """Add types to the dictionary by short name (class name with optional package alias)."""

        types_dict = {x.__name__: x for x in types}
        cls.get_type_dict().update(types_dict)

    @classmethod
    def get_type_by_short_name(cls, short_name: str) -> Type:
        """Get type from short name (class name with optional package alias)."""

        # Get dictionary of types indexed by alias
        type_dict_by_short_name = cls.get_type_dict()

        # TODO: Update to support short name with namespace prefix
        record_type = type_dict_by_short_name.get(short_name, None)
        if record_type is None:
            raise RuntimeError(
                f"Record class with short name {short_name} is not found "
                f"in the list of packages specified in settings."
            )
        return record_type

    @classmethod
    def get_type_dict(cls) -> Dict[str, Type]:
        """
        Get dictionary of types indexed by short name (class name with optional package alias).

        Notes:
            The result is returned in the alphabetical order of module.ClassName.
        """

        if cls._type_dict_by_short_name is None:
            # Load packages from Dynaconf
            context_settings = ContextSettings.instance()
            packages = context_settings.packages

            # Get modules for the specified packages
            modules = cls._get_modules(packages)

            # Get record types by iterating over modules
            record_types = set(
                record_type for module in modules for name, record_type in inspect.getmembers(module, is_key_or_record)
            )

            # Ensure names are unique
            # TODO: Support namespace aliases to resolve conflicts
            record_names = [record_type.__name__ for record_type in record_types]
            record_paths = [f"{record_type.__module__}.{record_type.__name__}" for record_type in record_types]

            # Check that there are no repeated names, report errors if there are
            if len(set(record_names)) != len(record_names):
                # Count the occurrences of each name in the list
                record_name_counts = Counter(record_names)

                # Find names that are repeated more than once
                repeated_names = [record_name for record_name, count in record_name_counts.items() if count > 1]

                # Report repeated names
                package_names_str = ", ".join(packages)
                repeated_names_str = ", ".join(repeated_names)
                raise RuntimeError(
                    f"The following class names in the list of packages {package_names_str} "
                    f"are repeated more than once: {repeated_names_str}"
                )

            # Create dictionary
            result = dict(zip(record_names, record_types))

            # Sort alphabetically by module_shortname.ClassName
            # TODO: Support module_shortname
            cls._type_dict_by_short_name = {key: result[key] for key in sorted(result)}

        return cls._type_dict_by_short_name

    @classmethod
    def for_key(cls, key: TypeDeclKey) -> Self:
        """Create or return cached object for the specified type declaration key."""
        class_path = f"{key.module.module_name}.{key.name}"
        return cls.for_class_path(class_path)

    @classmethod
    def for_class_path(cls, class_path: str) -> Dict[str, Dict]:
        """Create or return cached object for the specified class path in module.ClassName format."""

        record_type = ClassInfo.get_class_type(class_path)
        return cls.for_type(record_type)

    @classmethod
    @cached
    def for_type(cls, record_type: Type) -> Dict[str, Dict]:
        """
        Declarations for the specified type and all dependencies, returned as a dictionary.

        Args:
            record_type: Type of the record for which the schema is created.
        """
        dependencies = set()

        # Get or create type declaration the argument class
        type_decl_obj = TypeDecl.for_type(record_type, dependencies=dependencies)

        # Sort the list of dependencies
        # TODO: Use short name with alias
        dependencies = sorted(dependencies, key=lambda x: x.__name__)

        # Add to the dict by type name
        cls.add_types_to_dict_by_short_name(dependencies)

        # TODO: Restore after Enum decl generation is supported
        dependencies = [dependency_type for dependency_type in dependencies if not issubclass(dependency_type, Enum)]

        # Requested type is always first
        type_decl_list = [type_decl_obj] + [TypeDecl.for_type(dependency_type) for dependency_type in dependencies]

        # TODO: Move pascalize to a helper class
        result = {
            pascalize(f"{type_decl.module.module_name}.{type_decl.name}"): type_decl.to_type_decl_dict()
            for type_decl in type_decl_list
        }
        return result

    @classmethod
    @cached
    def _get_modules(cls, packages: List[str]) -> List[ModuleType]:
        """
        Get a list of ModuleType objects for submodules at all levels of the specified packages or root modules
        in the alphabetical order of dot-delimited module name.

        Args:
            packages: List of packages or root module strings in dot-delimited format, for example ['cl.runtime']
        """
        result = []
        for package in packages:
            # Import root module of the package
            root_module = importlib.import_module(package)
            result.append(root_module)  # Add the root module itself
            # Get module info for all submodules, note the trailing period added as per walk_packages documentation
            for module_info in walk_packages(root_module.__path__, root_module.__name__ + "."):
                module_name = module_info.name
                # Import the submodule using its full name
                submodule = importlib.import_module(module_name)
                result.append(submodule)

        # Sort the result by module path
        result = sorted(result, key=lambda module: module.__name__)
        return result

    @classmethod
    @cached
    def get_types_in_hierarchy(cls, record_type: Type) -> List[Type]:
        """
        Find all record types in hierarchy for given record_type.
        Include all base and child classes ordered by hierarchy.
        """

        get_key_type = getattr(record_type, "get_key_type", None)

        if get_key_type is None:
            raise RuntimeError(f"Type {record_type} is not record type.")

        # get key type
        # TODO (Roman): review get_key_type method of KeyProtocol. maybe it can be staticmethod or classmethod?
        key_type = get_key_type(None)

        # container to collect types
        types_ = defaultdict(list)

        for type_ in cls.get_types():
            # TODO (Roman): should not skip if input type is not key
            if type_ == record_type:
                # skip start type
                continue

            type_get_key_type = getattr(type_, "get_key_type", None)
            if type_get_key_type is None:
                # skip non-record types
                continue

            try:
                type_key_type = type_get_key_type(None)
            except AttributeError:
                # skip types whose key type unavailable from type
                continue

            if key_type != type_key_type:
                # skip types of another key type
                continue

            # try to resolve hierarchy
            try:
                i = type_.__mro__.index(key_type)
            except ValueError:
                # if key_type is not in mro default i = 1
                i = 1

            types_[i].append(type_)

        # final list of types in hierarchy
        result: List[Type] = []

        # order by place in hierarchy. more derived in the end.
        for k, v in sorted(types_.items(), key=lambda item: item[0]):
            result.extend(v)

        return result
