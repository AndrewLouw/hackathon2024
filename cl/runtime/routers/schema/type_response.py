# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations
import importlib
import inspect
from pkgutil import walk_packages
from types import ModuleType
from typing import List
from inflection import titleize
from pydantic import BaseModel
from cl.runtime.primitive.string_util import to_pascal_case


def is_record(cls):
    """Return true if the type is a record based on the presence of 'get_key' method."""
    return inspect.isclass(cls) and hasattr(cls, 'get_key') and callable(getattr(cls, 'get_key'))


class TypeResponse(BaseModel):
    """REST API response for a single item of the list returned by the /data/types route."""

    name: str
    """Class name (may be customized in settings)."""

    module: str
    """Module path in dot-delimited format (may be customized in settings)."""

    label: str
    """Type label displayed in the UI is humanized class name (may be customized in settings)."""

    class Config:
        alias_generator = to_pascal_case
        populate_by_name = True

    @staticmethod
    def get_modules(packages: List[str]) -> List[ModuleType]:
        """
        Get a list of ModuleType objects for submodules at all levels of the specified packages or root modules.
        Args:
            packages: List of packages or root module strings in dot-delimited format, for example ['cl.runtime']
        """
        result = []
        for package in packages:
            # Import root module of the package
            root_module = importlib.import_module(package)
            result.append(root_module)  # Add the root module itself
            # Get module info for all submodules, note the trailing period added as per walk_packages documentation
            for module_info in walk_packages(root_module.__path__, root_module.__name__ + "."):
                module_name = module_info.name
                # Import the submodule using its full name
                submodule = importlib.import_module(module_name)
                result.append(submodule)
        return result

    @staticmethod
    def get_types() -> List[TypeResponse]:
        """Implements /schema/types route."""

        packages = ["cl.runtime"]

        result = []
        modules = TypeResponse.get_modules(packages)
        record_types = [
            record_type
            for module in modules
            for name, record_type in inspect.getmembers(module, is_record)
        ]
        for record_type in record_types:
            type_response = TypeResponse(
                name=record_type.__name__,
                module=to_pascal_case(record_type.__module__),
                label=titleize(record_type.__name__),
            )
            result.append(type_response)
        return result
