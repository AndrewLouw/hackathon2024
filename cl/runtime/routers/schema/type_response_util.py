# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from cl.runtime.routers.schema.type_request import TypeRequest
from cl.runtime.schema.schema import Schema
from typing import Dict


class TypeResponseUtil:
    """Response helper class for the /schema/typeV2 route."""

    @staticmethod
    def get_type(request: TypeRequest) -> Dict[str, Dict]:
        """Implements /storage/get_datasets route."""

        # TODO: Check why empty module is passed, is module the short name prefix?
        record_type = Schema.get_type_by_short_name(request.name)
        result = Schema.for_type(record_type)

        for decl_name, decl_dict in result.items():
            # add Implement handlers block
            if declare_block := decl_dict.get("Declare"):
                if handlers_block := declare_block.get("Handlers"):
                    # TODO (Roman): skip abstract methods
                    implement_block = [{"Name": handler_decl.get("Name")} for handler_decl in handlers_block]
                    result[decl_name]["Implement"] = {"Handlers": implement_block}

        return result
