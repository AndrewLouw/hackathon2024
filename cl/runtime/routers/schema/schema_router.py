# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from typing import List, Dict
from fastapi import APIRouter, Header, Query

from cl.runtime.routers.schema.type_request import TypeRequest
from cl.runtime.routers.schema.type_response_util import TypeResponseUtil
from cl.runtime.routers.user_request import UserRequest
from cl.runtime.routers.schema.types_response_item import TypesResponseItem

TypesResponse = List[TypesResponseItem]
TypeResponse = Dict[str, int]

router = APIRouter()


@router.get("/types", response_model=TypesResponse)
async def get_types(user: str = Header(None, description="User identifier or identity token")) -> TypesResponse:
    """Information about the record types."""
    return TypesResponseItem.get_types(UserRequest(user=user))


@router.get("/typeV2", response_model=TypeResponse)
async def get_type(
        name: str = Query(..., description="Class name"),  # noqa Suppress report about shadowed built-in type
        module: str = Query(None, description="Dot-delimited module string"),
        user: str = Header(None, description="User identifier or identity token"),
) -> TypeResponse:
    """Schema for the specified type and its dependencies."""
    return TypeResponseUtil.get_type(TypeRequest(name=name, module=module, user=user))
