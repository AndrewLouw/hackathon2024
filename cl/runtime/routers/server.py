# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from cl.runtime.routers.auth import auth_router
from cl.runtime.routers.entity import entity_router
from cl.runtime.routers.health import health_router
from cl.runtime.routers.schema import schema_router
from cl.runtime.routers.storage import storage_router
from fastapi import FastAPI
from starlette.middleware.cors import CORSMiddleware

# Server
app = FastAPI()

origins = [
    "http://localhost",
    "http://localhost:8000",
    "http://localhost:7008",
    "http://localhost:3002",
]

app.add_middleware(
    CORSMiddleware,
    allow_origins=origins,
    allow_credentials=True,
    allow_methods=["*"],  # Specify allowed HTTP methods, e.g., ["GET", "POST"]
    allow_headers=["*"],  # Specify allowed headers, e.g., ["Content-Type", "Authorization"]
)


# Routers
app.include_router(health_router.router, prefix="", tags=["Health Check"])
app.include_router(auth_router.router, prefix="/auth", tags=["Authorization"])
app.include_router(schema_router.router, prefix="/schema", tags=["Schema"])
app.include_router(storage_router.router, prefix="/storage", tags=["Storage"])
app.include_router(entity_router.router, prefix="/entity", tags=["Entity"])
