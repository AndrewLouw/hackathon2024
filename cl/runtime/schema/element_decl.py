# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from cl.runtime.records.dataclasses.dataclass_mixin import datafield
from cl.runtime.schema.member_decl import MemberDecl
from dataclasses import dataclass
from typing import List
from typing import Optional


@dataclass(slots=True, kw_only=True)
class ElementDecl(MemberDecl):  # TODO: Consider renaming to TypeFieldDecl or FieldDecl
    """Type element declaration."""

    name: str = datafield()
    """Element name."""

    label: str | None = datafield()
    """Element label. If not specified, name is used instead."""

    comment: str | None = datafield()
    """Element comment. Contains addition information."""

    vector: Optional[bool] = datafield()  # TODO: Replace by container field with enum values vector/array, dict, DF
    """Flag indicating variable size array (vector) container."""

    optional: Optional[bool] = datafield()
    """Flag indicating optional element."""

    optional_vector_element: Optional[bool] = datafield()  # TODO: Rename to optional_element or optional_field
    """Flag indicating optional vector item element."""

    additive: Optional[bool] = datafield()
    """Optional flag indicating if the element is additive and that the total column can be shown in the UI."""

    format_: str | None = datafield(name="Format")  # TODO: Use Python interpolated string format
    """Specifies UI Format for the element."""

    alternate_of: str | None = datafield()
    """Link current element to AlternateOf element. In the editor these elements will be treated as a choice."""

