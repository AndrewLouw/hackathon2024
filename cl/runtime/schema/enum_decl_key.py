# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from cl.runtime.records.dataclasses.dataclass_data_mixin import datafield
from cl.runtime.records.dataclasses.dataclass_key_mixin import DataclassKeyMixin
from cl.runtime.schema.module_decl_key import ModuleDeclKey
from dataclasses import dataclass
from typing import Tuple


@dataclass(slots=True)
class EnumDeclKey(DataclassKeyMixin):
    """Enum declaration."""

    module: ModuleDeclKey = datafield()  # TODO: Merge with name
    """Module reference."""

    name: str = datafield()
    """Enum name is unique when combined with module."""

    def get_generic_key(self) -> Tuple:
        return EnumDeclKey, self.module.get_generic_key(), self.name
