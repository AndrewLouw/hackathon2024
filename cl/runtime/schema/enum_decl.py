# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from cl.runtime.records.dataclasses.dataclass_record_mixin import DataclassRecordMixin
from cl.runtime.records.dataclasses.dataclass_data_mixin import datafield
from cl.runtime.schema.enum_decl_key import EnumDeclKey
from cl.runtime.schema.enum_item_decl import EnumItemDecl
from dataclasses import dataclass
from typing import List


@dataclass(slots=True, kw_only=True)
class EnumDecl(EnumDeclKey, DataclassRecordMixin):
    """Enum declaration."""

    label: str | None = datafield()
    """Enum label."""

    comment: str | None = datafield()
    """Enum comment."""

    items: List[EnumItemDecl] = datafield()
    """Array of enum items."""

    def get_key(self) -> EnumDeclKey:
        return EnumDeclKey(module=self.module, name=self.name)

