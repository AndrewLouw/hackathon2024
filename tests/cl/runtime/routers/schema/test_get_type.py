# Copyright (C) 2023-present The Project Contributors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import pytest
from fastapi.testclient import TestClient
from cl.runtime.routers.server import app
from cl.runtime.routers.schema.schema_router import TypeResponse
from cl.runtime.routers.schema.type_response_util import TypeResponseUtil
from cl.runtime.routers.schema.type_request import TypeRequest

requests = [
    {"name": "StubClass"},
    {"name": "StubClass", "user": "TestUser"}
]

expected_result = {'Cl.Runtime.Backend.Core.StubClass': {'Module': {'ModuleName': 'Cl.Runtime.Backend.Core'}, 'Name': 'StubClass', 'Label': 'Ui App State', 'Comment': 'UiAppState.', 'DisplayKind': 'Basic', 'Elements': [{'Key': {'Module': {'ModuleName': 'Cl.Runtime.Backend.Core'}, 'Name': 'User'}, 'Name': 'User', 'Comment': 'A user the app state is applied for.', 'Optional': True}, {'Data': {'Module': {'ModuleName': 'Cl.Runtime.Backend.Core'}, 'Name': 'TabInfo'}, 'Name': 'OpenedTabs', 'Comment': 'Information about opened tabs.', 'Vector': True, 'Optional': True}, {'Value': {'Type': 'Int'}, 'Name': 'ActiveTabIndex', 'Comment': 'Index of active opened tab.', 'Optional': True}, {'Value': {'Type': 'Dict'}, 'Name': 'Versions', 'Comment': 'Component versions.', 'Optional': True}, {'Value': {'Type': 'String'}, 'Name': 'BackendVersion', 'Comment': 'DEPRECATED. Use versions instead.', 'Optional': True}, {'Value': {'Type': 'String'}, 'Name': 'ApplicationName', 'Comment': 'Application name.', 'Optional': True}, {'Value': {'Type': 'Bool'}, 'Name': 'ReadOnly', 'Comment': 'Flag indicating that UI is read-only.', 'Optional': True}, {'Enum': {'Module': {'ModuleName': 'Cl.Runtime.Backend.Core'}, 'Name': 'AppTheme'}, 'Name': 'ApplicationTheme', 'Comment': 'Application theme (dark, light, etc.).', 'Optional': True}], 'Keys': ['User'], 'Implement': {'Handlers': []}}, 'Cl.Runtime.Backend.Core.User': {'Module': {'ModuleName': 'Cl.Runtime.Backend.Core'}, 'Name': 'User', 'Label': 'User', 'Comment': 'User which is allowed to log in.', 'DisplayKind': 'Basic', 'Elements': [{'Value': {'Type': 'String'}, 'Name': 'Username'}, {'Value': {'Type': 'String'}, 'Name': 'FirstName', 'Comment': 'First name of the user.'}, {'Value': {'Type': 'String'}, 'Name': 'LastName', 'Comment': 'Last name of the user.'}, {'Value': {'Type': 'String'}, 'Name': 'Email', 'Comment': 'Email of the user.', 'Optional': True}], 'Keys': ['Username'], 'Implement': {'Handlers': []}}, 'Cl.Runtime.Backend.Core.TabInfo': {'Module': {'ModuleName': 'Cl.Runtime.Backend.Core'}, 'Name': 'TabInfo', 'Label': 'Tab Info', 'Comment': 'Tab info.', 'Kind': 'Element', 'DisplayKind': 'Basic', 'Elements': [{'Data': {'Module': {'ModuleName': 'Cl.Runtime.Backend.Core'}, 'Name': 'BaseTypeInfo'}, 'Name': 'Type', 'Comment': 'Type.'}, {'Value': {'Type': 'Key'}, 'Name': 'Key', 'Comment': 'Key.', 'Optional': True}], 'Keys': None, 'Implement': {'Handlers': []}}, 'Cl.Runtime.Backend.Core.BaseTypeInfo': {'Module': {'ModuleName': 'Cl.Runtime.Backend.Core'}, 'Name': 'BaseTypeInfo', 'Label': 'Base Type Info', 'Comment': 'Base type info.', 'Kind': 'Element', 'DisplayKind': 'Basic', 'Elements': [{'Value': {'Type': 'String'}, 'Name': 'Name', 'Comment': 'Name of type.'}, {'Value': {'Type': 'String'}, 'Name': 'Module', 'Comment': 'Module of type.'}, {'Value': {'Type': 'String'}, 'Name': 'Label', 'Comment': 'Label of type.'}], 'Keys': None, 'Implement': {'Handlers': []}}}


def test_method():
    """Test coroutine for /schema/typeV2 route."""

    for request in requests:

        # Run the coroutine wrapper added by the FastAPI decorator and get the result
        request_obj = TypeRequest(**request)
        result = TypeResponseUtil.get_type(request_obj)

        # Check result
        assert result == expected_result


def test_api():
    """Test REST API for /schema/typeV2 route."""

    with TestClient(app) as client:
        for request in requests:

            # Split request headers and query
            request_headers = {"user": request.get("user")}
            request_params = {"name": request.get("name"), "module": request.get("module")}

            # Eliminate empty keys
            request_headers = {k: v for k, v in request_headers.items() if v is not None}
            request_params = {k: v for k, v in request_params.items() if v is not None}

            # Get response
            response = client.get("/schema/typeV2", headers=request_headers, params=request_params)
            assert response.status_code == 200
            result = response.json()

            # Check result
            assert result == expected_result


if __name__ == "__main__":
    pytest.main([__file__])
