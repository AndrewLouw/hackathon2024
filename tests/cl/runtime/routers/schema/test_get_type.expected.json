{
    "Cl.Runtime.Backend.Core.UiAppState.UiAppState": {
        "Module": {
            "ModuleName": "Cl.Runtime.Backend.Core.UiAppState"
        },
        "Name": "UiAppState",
        "Label": "Ui App State",
        "Comment": "UiAppState.",
        "DisplayKind": "Basic",
        "Elements": [
            {
                "Key": {
                    "Module": {
                        "ModuleName": "Cl.Runtime.Backend.Core.User"
                    },
                    "Name": "User"
                },
                "Name": "User",
                "Comment": "A user the app state is applied for."
            },
            {
                "Data": {
                    "Module": {
                        "ModuleName": "Cl.Runtime.Backend.Core.TabInfo"
                    },
                    "Name": "TabInfo"
                },
                "Name": "OpenedTabs",
                "Comment": "Information about opened tabs.",
                "Vector": true,
                "Optional": true
            },
            {
                "Value": {
                    "Type": "Int"
                },
                "Name": "ActiveTabIndex",
                "Comment": "Index of active opened tab.",
                "Optional": true
            },
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "BackendVersion",
                "Comment": "DEPRECATED. Use versions instead.",
                "Optional": true
            },
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "ApplicationName",
                "Comment": "Application name.",
                "Optional": true
            },
            {
                "Value": {
                    "Type": "Bool"
                },
                "Name": "ReadOnly",
                "Comment": "Flag indicating that UI is read-only.",
                "Optional": true
            },
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "ApplicationTheme",
                "Comment": "Application theme (dark, light, etc.).",
                "Optional": true
            }
        ],
        "Keys": [
            "User"
        ]
    },
    "Cl.Runtime.Backend.Core.User.User": {
        "Module": {
            "ModuleName": "Cl.Runtime.Backend.Core.User"
        },
        "Name": "User",
        "Label": "User",
        "Comment": "User which is allowed to log in.",
        "DisplayKind": "Basic",
        "Elements": [
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "Username"
            },
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "FirstName",
                "Comment": "First name of the user."
            },
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "LastName",
                "Comment": "Last name of the user."
            },
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "Email",
                "Comment": "Email of the user.",
                "Optional": true
            }
        ],
        "Keys": [
            "Username"
        ]
    },
    "Cl.Runtime.Backend.Core.TabInfo.TabInfo": {
        "Module": {
            "ModuleName": "Cl.Runtime.Backend.Core.TabInfo"
        },
        "Name": "TabInfo",
        "Label": "Tab Info",
        "Comment": "Tab info.",
        "Kind": "Element",
        "DisplayKind": "Basic",
        "Elements": [
            {
                "Data": {
                    "Module": {
                        "ModuleName": "Cl.Runtime.Backend.Core.BaseTypeInfo"
                    },
                    "Name": "BaseTypeInfo"
                },
                "Name": "Type",
                "Comment": "Type."
            },
            {
                "Value": {
                    "Type": "Key"
                },
                "Name": "Key",
                "Comment": "Key.",
                "Optional": true
            }
        ],
        "Keys": null
    },
    "Cl.Runtime.Backend.Core.BaseTypeInfo.BaseTypeInfo": {
        "Module": {
            "ModuleName": "Cl.Runtime.Backend.Core.BaseTypeInfo"
        },
        "Name": "BaseTypeInfo",
        "Label": "Base Type Info",
        "Comment": "Base type info.",
        "Kind": "Element",
        "DisplayKind": "Basic",
        "Elements": [
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "Name",
                "Comment": "Name of type."
            },
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "Module",
                "Comment": "Module of type."
            },
            {
                "Value": {
                    "Type": "String"
                },
                "Name": "Label",
                "Comment": "Label of type."
            }
        ],
        "Keys": null
    }
}
